### PR TITLE
[CWS] Use ring buffers instead of perf maps when available

### PR DESCRIPTION
**Pull request by lebauce**
### Reviewer's Checklist

- [ ] If known, an appropriate milestone has been selected; otherwise the `Triage` milestone is set.
- [ ] Use the `major_change` label if your change either has a major impact on the code base, is impacting multiple teams or is changing important well-established internals of the Agent. This label will be use during QA to make sure each team pay extra attention to the changed behavior. For any customer facing change use a releasenote.
- [ ] A [release note](https://github.com/DataDog/datadog-agent/blob/main/docs/dev/contributing.md#reno) has been added or the `changelog/no-changelog` label has been applied.
- [ ] Changed code has automated tests for its functionality.
- [ ] Adequate QA/testing plan information is provided if the `qa/skip-qa` label is not applied.
- [ ] At least one `team/..` label has been applied, indicating the team(s) that should QA this change.
- [ ] If applicable, docs team has been notified or [an issue has been opened on the documentation repo](https://github.com/DataDog/documentation/issues/new).
- [ ] If applicable, the `need-change/operator` and `need-change/helm` labels have been applied.
- [ ] If applicable, the [config template](https://github.com/DataDog/datadog-agent/blob/main/pkg/config/config_template.yaml) has been updated.

### CODE REVIEW

File: pkg/ebpf/c/bpf_helpers.h [hunk 1]
```diff
@@ -44,6 +44,7 @@ static unsigned long long (*bpf_get_current_pid_tgid)(void) = (void*)BPF_FUNC_ge
 static unsigned long long (*bpf_get_current_uid_gid)(void) = (void*)BPF_FUNC_get_current_uid_gid;
 static int (*bpf_get_current_comm)(void* buf, int buf_size) = (void*)BPF_FUNC_get_current_comm;
 static int (*bpf_perf_event_read)(void* map, int index) = (void*)BPF_FUNC_perf_event_read;
+static int (*bpf_ringbuf_output)(void *ringbuf, void *data, u64 size, u64 flags) = (void*)BPF_FUNC_ringbuf_output;
```
Comment (paulcacheux) [inline pkg/ebpf/c/bpf_helpers.h:47]: You need to gate this with LINUX_VERSION_CODE #if

**Review by djmitche — APPROVED** [review]
👍🏽 pkg/status/templates/runtimesecurity.tmpl

**Review by paulcacheux — CHANGES_REQUESTED** [review]

**Review by paulcacheux — APPROVED** [review]

File: pkg/security/ebpf/manager.go [hunk 2]
```diff
@@ -53,10 +55,15 @@ func NewDefaultOptions() manager.Options {
...
-	return &manager.Manager{
-		Probes:   probes.AllProbes(),
-		Maps:     probes.AllMaps(),
-		PerfMaps: probes.AllPerfMaps(),
+func NewRuntimeSecurityManager(supportsRingBuffers bool) *manager.Manager {
+	manager := &manager.Manager{
+		Probes: probes.AllProbes(),
+		Maps:   probes.AllMaps(),
+	}
+	if supportsRingBuffers {
```
Comment (paulcacheux) [inline pkg/security/ebpf/manager.go:63]: Maybe we should have a config option to force the usage of the perf map, even if we support the ring buffer ?

File: pkg/security/ebpf/manager.go [hunk 3]
```diff
@@ -30,6 +30,8 @@ func NewDefaultOptions() manager.Options {
 		// following form: (1 + 2^n) * pages. Checkout https://github.com/DataDog/ebpf for more.
 		DefaultPerfRingBufferSize: 4097 * os.Getpagesize(),
 
+		DefaultRingBufferSize: 4097 * os.Getpagesize(),
```
Comment (paulcacheux) [inline pkg/security/ebpf/manager.go:33]:
```suggestion
		DefaultRingBufferSize: 4096 * os.Getpagesize(),
```

File: pkg/security/ebpf/probes/all.go [hunk 4]
```diff
@@ -177,6 +179,13 @@ func AllMapSpecEditors(numCPU int, tracedCgroupsCount int, cgroupWaitListSize in
 			EditorFlag: manager.EditFlags,
 		}
 	}
+	if supportsRingBuffers {
+		editors["events"] = manager.MapSpecEditor{
+			MaxEntries: uint32(4096 * os.Getpagesize()),
```
Comment (paulcacheux) [inline pkg/security/ebpf/probes/all.go:184]: Maybe use the default constant ?

File: pkg/security/ebpf/manager.go (hunk 2 above)
Comment (lebauce) [reply pkg/security/ebpf/manager.go:58]: Done

File: tasks/system_probe.py [hunk 5]
```diff
@@ -693,12 +697,12 @@ def build_security_probe_ebpf_files(ctx, build_dir, parallel_build=True):
...
-def build_security_ebpf_files(ctx, build_dir, parallel_build=True):
-    build_security_probe_ebpf_files(ctx, build_dir, parallel_build)
-    build_security_offset_guesser_ebpf_files(ctx, build_dir)
+def build_security_ebpf_files(ctx, build_dir, parallel_build=True, kernel_release=''):
+    build_security_probe_ebpf_files(ctx, build_dir, parallel_build, kernel_release=kernel_release)
+    build_security_offset_guesser_ebpf_files(ctx, build_dir, kernel_release=kernel_release)
 
 
-def build_object_files(ctx, parallel_build):
+def build_object_files(ctx, parallel_build, kernel_release=''):
```
Comment (brycekahle) [inline tasks/system_probe.py:705]:
```suggestion
def build_object_files(ctx, parallel_build, kernel_release=None):
```

File: tasks/system_probe.py [hunk 6]
```diff
@@ -693,12 +697,12 @@ def build_security_probe_ebpf_files(ctx, build_dir, parallel_build=True):
             p.join()
 
 
-def build_security_ebpf_files(ctx, build_dir, parallel_build=True):
-    build_security_probe_ebpf_files(ctx, build_dir, parallel_build)
-    build_security_offset_guesser_ebpf_files(ctx, build_dir)
+def build_security_ebpf_files(ctx, build_dir, parallel_build=True, kernel_release=''):
```
Comment (brycekahle) [inline tasks/system_probe.py:700]:
```suggestion
def build_security_ebpf_files(ctx, build_dir, parallel_build=True, kernel_release=None):
```

File: tasks/system_probe.py [hunk 7]
```diff
@@ -402,18 +403,21 @@ def get_ebpf_targets():
     return files
 
 
-def get_linux_header_dirs():
-    os_info = os.uname()
+def get_linux_header_dirs(kernel_release=''):
```
Comment (brycekahle) [inline tasks/system_probe.py:406]:
```suggestion
def get_linux_header_dirs(kernel_release=None):
```

File: tasks/system_probe.py [hunk 8]
```diff
@@ -465,7 +469,7 @@ def get_linux_header_dirs():
     return dirs
 
 
-def get_ebpf_build_flags(target=None):
+def get_ebpf_build_flags(target=None, kernel_release=''):
```
Comment (brycekahle) [inline tasks/system_probe.py:472]:
```suggestion
def get_ebpf_build_flags(target=None, kernel_release=None):
```

File: tasks/system_probe.py [hunk 9]
```diff
@@ -60,6 +60,7 @@ def build(
     nikos_embedded_path=None,
     bundle_ebpf=False,
     parallel_build=True,
+    kernel_release='',
```
Comment (brycekahle) [inline tasks/system_probe.py:63]:
```suggestion
    kernel_release=None,
```

File: pkg/security/ebpf/kernel/kernel.go [hunk 10]
```diff
@@ -244,3 +249,46 @@ func (k *Version) IsAmazonLinuxKernel() bool {
...
+	*k.haveMmapableMaps = err == nil
+
+	if err != nil {
+		return false
+	}
+	_ = m.Close()
+	return true
+}
+
+func (k *Version) HaveRingBuffers() bool {
```
Comment (brycekahle) [inline pkg/security/ebpf/kernel/kernel.go:276]: Did you see the `cilium/ebpf` functions for this? You can do `features.HaveMapType(ebpf.RingBuf)`

Comment (lebauce) [reply pkg/security/ebpf/kernel/kernel.go:276]: The function was not exported at the time. Will use it

File: .gitlab-ci.yml [hunk 11]
```diff
@@ -139,8 +139,8 @@ variables:
   DATADOG_AGENT_BUILDERS: v8333476-cd1c34b
   DATADOG_AGENT_WINBUILDIMAGES: v8334078-66730ef
   DATADOG_AGENT_ARMBUILDIMAGES: v8334078-66730ef
-  DATADOG_AGENT_SYSPROBE_BUILDIMAGES: v8334078-66730ef
-  DATADOG_AGENT_NIKOS_BUILDIMAGES: v8334078-66730ef
+  DATADOG_AGENT_SYSPROBE_BUILDIMAGES:  v8364396-f91ff4f
```
[the remark this answers is not in the document]
Comment (lebauce) [reply .gitlab-ci.yml:142]: Nice catch

**Review by safchain — APPROVED** [review]

**Review by nplanel — APPROVED** [review]

**Pull request merged**